src/main/java/net/spy/memcached/protocol/ascii/CollectionBulkInsertOperationImpl.java

```diff
@@ -104,6 +118,12 @@ assert getState() == OperationState.READING
         cb.gotStatus(insert.getKey(index), status);
         successAll = false;
       }
+      /* ENABLE_MIGRATION if */
+      if (needRedirect()) {
+        transitionState(OperationState.REDIRECT);
+        return;
+      }
+      /* ENABLE_MIGRATION end */
```

Comment:
bulk 연산 중 smget과 bulk insert에 대해 코멘트 남기고,
다른 bulk 연산들은 확인해 보지 않았습니다.
아마 비슷한 형태가 아닐까 싶어 여기서 리뷰는 중지합니다.
수정되고 나서 다시 살펴보겠습니다.

Reply:
hasNotMyKey() 조건은 if (isNotPiped()) 문 안에도, else 문 안에도 있어야 하는 부분이므로 /* ENABLE_REPLICATION end */ 아래에 두었습니다.

아래의 hasNotMyKey() 조건과 isNotPiped() 조건을 모두 만족하는 경우에 setNextOpIndex() 메소드를 호출 할지 말지 고민이 되는데요. isNotPiped() 조건을 만족하기 때문에 index == 0이고 nextOpIndex의 초기값도 0이므로 사실 호출을 안 해도 됩니다. 그런데 '이 index부터 redirect 된다'는 표현을 명시할 수 있습니다. setNextOpIndex() 메소드 호출은 빠져도 좋을까요?

```java
    /* ENABLE_MIGRATION if */
    if (hasNotMyKey(line)) {
      if (insert.isNotPiped()) {
        insert.setNextOpIndex(index); // FIXME: Is it needed to be called for better readability?
        addRedirectSingleKeyOperation(line, insert.getKey(index));
        transitionState(OperationState.REDIRECT);
      } else {
        // NOT need to call setNextOpIndex()
        // because multi key operation will be cloned with redirect keys.
        // nextOpIndex will be discarded when operation is cloned.
        addRedirectMultiKeyOperation(line, insert.getKey(index));
        index++;
      }
      return;
    }
    /* ENABLE_MIGRATION end */
```

Reply:
- addRedirectMultiKeyOperation() 함수를 호출하시죠.
   - key는 1개이지만 연산 자체가 bulk insert 성격이기 때문입니다.
   - 내부적으로 clone하여 처리하는 cost는 가집니다.
   - setNextOpIndex() 호출하지 않아도 됩니다.
- 앞서 코멘트한 대로, 원래 로직 안에 hasNotMyKey() case를 추가하는 형태가 좋겠습니다.
   - 원래 로직의 의미를 이해하면서 NOT_MY_KEY 처리 코드를 볼 수 있어 이해하기 쉽습니다.
   - redirect needed 로직도 원래 로직 안에 있으므로, 이와도 일관성이 있게 됩니다.

Reply:
<details>
<summary>공통 로직으로 처리하는 구현</summary>

```java
  @Override
  public void handleLine(String line) {
    assert getState() == OperationState.READING
            : "Read ``" + line + "'' when in " + getState() + " state";
    /* ENABLE_REPLICATION if */
    if (hasSwitchedOver(line)) {
      this.insert.setNextOpIndex(index);
      receivedMoveOperations(line);
      return;
    }
    /* ENABLE_REPLICATION end */
    /* ENABLE_MIGRATION if */
    if (hasNotMyKey(line)) {
      // NOT need to call setNextOpIndex()
      // because multi key operation will be cloned with redirect keys.
      // nextOpIndex will be discarded when operation is cloned.
      addRedirectMultiKeyOperation(line, insert.getKey(index));
      if (insert.isNotPiped()) {
        transitionState(OperationState.REDIRECT);
        return;
      } else {
        index++;
      }
    }
    /* ENABLE_MIGRATION end */
    if (insert.isNotPiped()) {
      OperationStatus status = matchStatus(line, STORED, CREATED_STORED,
              NOT_FOUND, ELEMENT_EXISTS, OVERFLOWED, OUT_OF_RANGE,
              TYPE_MISMATCH, BKEY_MISMATCH);
      if (!status.isSuccess()) {
        cb.gotStatus(insert.getKey(index), status);
        successAll = false;
      }
      cb.receivedStatus((successAll) ? END : FAILED_END);
      transitionState(OperationState.COMPLETE);
      return;
    }

    /*
      RESPONSE <count>\r\n
      <status of the 1st pipelined command>\r\n
      [ ... ]
      <status of the last pipelined command>\r\n
      END|PIPE_ERROR <error_string>\r\n
    */
    if (line.startsWith("END") || line.startsWith("PIPE_ERROR ")) {
      /* ENABLE_MIGRATION if */
      if (needRedirect()) {
        transitionState(OperationState.REDIRECT);
        return;
      }
      /* ENABLE_MIGRATION end */
      cb.receivedStatus((successAll) ? END : FAILED_END);
      transitionState(OperationState.COMPLETE);
    } else if (line.startsWith("RESPONSE ")) {
      getLogger().debug("Got line %s", line);

      // TODO server should be fixed
      line = line.replace("   ", " ");
      line = line.replace("  ", " ");

      String[] stuff = line.split(" ");
      assert "RESPONSE".equals(stuff[0]);
      count = Integer.parseInt(stuff[1]);
    } else {
      OperationStatus status = matchStatus(line, STORED, CREATED_STORED,
              NOT_FOUND, ELEMENT_EXISTS, OVERFLOWED, OUT_OF_RANGE,
              TYPE_MISMATCH, BKEY_MISMATCH);

      if (!status.isSuccess()) {
        cb.gotStatus(insert.getKey(index), status);
        successAll = false;
      }

      index++;
    }
  }
```

</details>

<details>
<summary>기존 로직에 추가하는 구현</summary>

```java
  @Override
  public void handleLine(String line) {
    assert getState() == OperationState.READING
            : "Read ``" + line + "'' when in " + getState() + " state";
    /* ENABLE_REPLICATION if */
    if (hasSwitchedOver(line)) {
      this.insert.setNextOpIndex(index);
      receivedMoveOperations(line);
      return;
    }
    /* ENABLE_REPLICATION end */
    if (insert.isNotPiped()) {
      /* ENABLE_MIGRATION if */
      if (hasNotMyKey(line)) {
        addRedirectMultiKeyOperation(line);
        transitionState(OperationState.REDIRECT);
        return;
      }
      /* ENABLE_MIGRATION end */
      OperationStatus status = matchStatus(line, STORED, CREATED_STORED,
              NOT_FOUND, ELEMENT_EXISTS, OVERFLOWED, OUT_OF_RANGE,
              TYPE_MISMATCH, BKEY_MISMATCH);
      if (!status.isSuccess()) {
        cb.gotStatus(insert.getKey(index), status);
        successAll = false;
      }
      cb.receivedStatus((successAll) ? END : FAILED_END);
      transitionState(OperationState.COMPLETE);
      return;
    }

    /*
      RESPONSE <count>\r\n
      <status of the 1st pipelined command>\r\n
      [ ... ]
      <status of the last pipelined command>\r\n
      END|PIPE_ERROR <error_string>\r\n
    */
    if (line.startsWith("END") || line.startsWith("PIPE_ERROR ")) {
      /* ENABLE_MIGRATION if */
      if (needRedirect()) {
        transitionState(OperationState.REDIRECT);
        return;
      }
      /* ENABLE_MIGRATION end */
      cb.receivedStatus((successAll) ? END : FAILED_END);
      transitionState(OperationState.COMPLETE);
    } else if (line.startsWith("RESPONSE ")) {
      getLogger().debug("Got line %s", line);

      // TODO server should be fixed
      line = line.replace("   ", " ");
      line = line.replace("  ", " ");

      String[] stuff = line.split(" ");
      assert "RESPONSE".equals(stuff[0]);
      count = Integer.parseInt(stuff[1]);
      /* ENABLE_MIGRATION if */
    } else if (hasNotMyKey(line)) {
      addRedirectMultiKeyOperation(line);
      index++;
      /* ENABLE_MIGRATION end */
    } else {
      OperationStatus status = matchStatus(line, STORED, CREATED_STORED,
              NOT_FOUND, ELEMENT_EXISTS, OVERFLOWED, OUT_OF_RANGE,
              TYPE_MISMATCH, BKEY_MISMATCH);

      if (!status.isSuccess()) {
        cb.gotStatus(insert.getKey(index), status);
        successAll = false;
      }

      index++;
    }
  }

  private void addRedirectMultiKeyOperation(String line) {
    // NOT need to call setNextOpIndex()
    // because multi key operation will be cloned with redirect keys.
    // nextOpIndex will be discarded when operation is cloned.
    addRedirectMultiKeyOperation(line, insert.getKey(index));
  }
```

</details>

반영하기 전, 우선 두 방식의 구현 예제를 작성했습니다. 현재는 모든 Operation 클래스에서 전자의 방식을 채택하고 있는데, 후자의 방식이 더 낫다면 다른 Operation 클래스에도 후자의 방식으로 수정하고자 합니다.

Reply:
nextOpIndex 설정은 replication 처리에서만 사용하는 개념이므로, 아래 주석은 없어도 될 것 같습니다. 
```java
      // NOT need to call setNextOpIndex()
      // because multi key operation will be cloned with redirect keys.
      // nextOpIndex will be discarded when operation is cloned.
```
그러면, 기존대로 아래의 코드를 넣어 두시죠. 
`return;` 수행 위치가 변경되었음을 참고하세요.

```java
    /* ENABLE_MIGRATION if */
    if (hasNotMyKey(line)) {
      addRedirectMultiKeyOperation(line, insert.getKey(index));
      if (insert.isNotPiped()) {
        transitionState(OperationState.REDIRECT);
      } else {
        index++;
      }
      return;
    }
    /* ENABLE_MIGRATION end */
```

그리고, 아래 응답 처리하는 코드의 구현 방식은 나중에 변경될 수 있음을 참고해 주세요.
- NOT_MY_KEY
- SWITCHOVER, REPL_SLAVE

Reply:
그리고, 현재 PR에 코멘트가 많다 보니, 반응 속도가 느립니다.
현재 PR은 그대로 두고, 새로운 PR로 이번에 수정할 사항을 올려주고 review discussion을 계속 진행합시다.